File: backend/app/portfolio/rebalance.py
"""Rebalancing: compare current vs target asset allocation, compute drift and suggested moves.

Two modes:
  * ``rebalance``  — assumes you can buy AND sell to hit targets exactly.
  * ``new_money``  — only deploy fresh money into underweight buckets (no selling), which is
                     how most retail investors actually rebalance and avoids tax/exit-load events.

This produces *informational* drift math and generic, non-personalized tax/exit-load reminders.
It is explicitly NOT financial advice.
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.models import TargetAllocation
from app.domain.taxonomy import ASSET_CLASS_ORDER, AssetClass
from app.portfolio.aggregate import summary

# Generic, non-personalized reminders surfaced when a bucket would be trimmed.
_SELL_WARNINGS = {
    AssetClass.EQUITY.value: "Equity/equity-fund sales: LTCG (>1yr) taxed 12.5% above ₹1.25L/yr; "
                             "STCG (<1yr) at 20%. Equity funds may levy ~1% exit load if <1yr.",
    AssetClass.INTERNATIONAL_EQUITY.value: "International funds are taxed like debt/non-equity for "
                             "many schemes — confirm the scheme's tax treatment and holding period.",
    AssetClass.DEBT.value: "Debt fund gains are taxed at your slab rate (post-Apr-2023 units). "
                           "Check exit load on short-duration/credit funds.",
    AssetClass.GOLD.value: "Gold funds/ETFs: gains taxable; SGBs are illiquid before maturity and "
                           "best redeemed at the RBI window/exchange.",
}


def _targets(db: Session) -> dict[str, float]:
    rows = db.scalars(
        select(TargetAllocation).where(TargetAllocation.level == "asset_class")
    ).all()
    return {r.bucket: r.target_pct for r in rows}


def rebalance_plan(db: Session, mode: str = "rebalance", new_money: float = 0.0) -> dict:
    s = summary(db)
    net_worth = s["net_worth"]
    current_by_class = {d["label"]: d for d in s["asset_allocation"]}
    targets = _targets(db)

    buckets = list(dict.fromkeys(
        [a.value for a in ASSET_CLASS_ORDER]
        + list(current_by_class.keys())
        + list(targets.keys())
    ))

    new_total = net_worth + (new_money if mode == "new_money" else 0.0)
    lines = []
    for bucket in buckets:
        cur = current_by_class.get(bucket)
        cur_value = cur["value"] if cur else 0.0
        cur_pct = cur["pct"] if cur else 0.0
        target_pct = targets.get(bucket)
        if target_pct is None and cur_value == 0:
            continue

        target_pct = target_pct or 0.0
        target_value = round(target_pct / 100 * new_total, 2)
        drift_pct = round(cur_pct - target_pct, 2)
        drift_value = round(cur_value - target_value, 2)

        if mode == "new_money":
            # Only top up underweight buckets; never suggest selling.
            suggested = round(max(0.0, target_value - cur_value), 2)
            action = "buy" if suggested > 0 else "hold"
            amount = suggested
        else:
            action = "sell" if drift_value > 0 else ("buy" if drift_value < 0 else "hold")
            amount = round(abs(drift_value), 2)

        warnings = []
        if action == "sell":
            w = _SELL_WARNINGS.get(bucket)
            if w:
                warnings.append(w)

        lines.append({
            "bucket": bucket,
            "current_value": round(cur_value, 2),
            "current_pct": round(cur_pct, 2),
            "target_pct": round(target_pct, 2),
            "target_value": target_value,
            "drift_pct": drift_pct,
            "drift_value": drift_value,
            "action": action,
            "amount": amount,
            "status": "overweight" if drift_value > 1 else ("underweight" if drift_value < -1 else "on target"),
            "warnings": warnings,
        })

    # In new-money mode, scale buys so they sum to the injected amount (proportional to need).
    if mode == "new_money" and new_money > 0:
        total_buy = sum(l["amount"] for l in lines if l["action"] == "buy")
        if total_buy > 0:
            scale = new_money / total_buy
            for l in lines:
                if l["action"] == "buy":
                    l["amount"] = round(l["amount"] * scale, 2)

    has_targets = bool(targets)
    return {
        "mode": mode,
        "new_money": new_money,
        "net_worth": net_worth,
        "has_targets": has_targets,
        "targets_sum": round(sum(targets.values()), 2),
        "lines": lines,
        "disclaimer": "Informational drift analysis only — not personalized investment advice. "
                      "Verify tax, holding period and exit loads before acting.",
    }

File: frontend/src/engine/portfolio/rebalance.ts
// Rebalancing — TS mirror of backend/app/portfolio/rebalance.py.

import { Store } from "../store/store";
import { ASSET_CLASS_ORDER, AssetClass } from "../taxonomy";
import { summary } from "./aggregate";

const round = (n: number) => Math.round(n * 100) / 100;

const SELL_WARNINGS: Record<string, string> = {
  [AssetClass.EQUITY]: "Equity/equity-fund sales: LTCG (>1yr) taxed 12.5% above ₹1.25L/yr; STCG (<1yr) at 20%. Equity funds may levy ~1% exit load if <1yr.",
  [AssetClass.INTERNATIONAL_EQUITY]: "International funds are taxed like debt/non-equity for many schemes — confirm the scheme's tax treatment and holding period.",
  [AssetClass.DEBT]: "Debt fund gains are taxed at your slab rate (post-Apr-2023 units). Check exit load on short-duration/credit funds.",
  [AssetClass.GOLD]: "Gold funds/ETFs: gains taxable; SGBs are illiquid before maturity and best redeemed at the RBI window/exchange.",
};

function targetsMap(store: Store): Record<string, number> {
  const out: Record<string, number> = {};
  for (const t of store.targets) if (t.level === "asset_class") out[t.bucket] = t.target_pct;
  return out;
}

export function rebalancePlan(store: Store, mode: "rebalance" | "new_money" = "rebalance", newMoney = 0): any {
  const s = summary(store);
  const netWorth = s.net_worth;
  const currentByClass = new Map<string, { value: number; pct: number }>(
    s.asset_allocation.map((d: any) => [d.label, d])
  );
  const targets = targetsMap(store);

  const buckets = [...new Set([...ASSET_CLASS_ORDER, ...currentByClass.keys(), ...Object.keys(targets)])];
  const newTotal = netWorth + (mode === "new_money" ? newMoney : 0);
  const lines: any[] = [];

  for (const bucket of buckets) {
    const cur = currentByClass.get(bucket);
    const curValue = cur ? cur.value : 0;
    const curPct = cur ? cur.pct : 0;
    let targetPct = targets[bucket];
    if (targetPct === undefined && curValue === 0) continue;
    targetPct = targetPct || 0;
    const targetValue = round((targetPct / 100) * newTotal);
    const driftPct = round(curPct - targetPct);
    const driftValue = round(curValue - targetValue);

    let action: string;
    let amount: number;
    if (mode === "new_money") {
      const suggested = round(Math.max(0, targetValue - curValue));
      action = suggested > 0 ? "buy" : "hold";
      amount = suggested;
    } else {
      action = driftValue > 0 ? "sell" : driftValue < 0 ? "buy" : "hold";
      amount = round(Math.abs(driftValue));
    }
    const warnings: string[] = [];
    if (action === "sell" && SELL_WARNINGS[bucket]) warnings.push(SELL_WARNINGS[bucket]);

    lines.push({
      bucket, current_value: round(curValue), current_pct: round(curPct),
      target_pct: round(targetPct), target_value: targetValue, drift_pct: driftPct, drift_value: driftValue,
      action, amount,
      status: driftValue > 1 ? "overweight" : driftValue < -1 ? "underweight" : "on target",
      warnings,
    });
  }

  if (mode === "new_money" && newMoney > 0) {
    const totalBuy = lines.filter((l) => l.action === "buy").reduce((a, l) => a + l.amount, 0);
    if (totalBuy > 0) {
      const scale = newMoney / totalBuy;
      for (const l of lines) if (l.action === "buy") l.amount = round(l.amount * scale);
    }
  }

  return {
    mode, new_money: newMoney, net_worth: netWorth,
    has_targets: Object.keys(targets).length > 0,
    targets_sum: round(Object.values(targets).reduce((a, b) => a + b, 0)),
    lines,
    disclaimer: "Informational drift analysis only — not personalized investment advice. Verify tax, holding period and exit loads before acting.",
  };
}
